Update with in game admin

gartnera opened the pull request:
I used tinyctf last year for a CTF competition for my school. I really like the simplicity of code that Flask provides.

Anyway, I decided to add an in game admin function. The first user that registers is the admin. They can add categories and add/edit tasks (with file upload).

I've also added a bunch of other things (start time, IP logging, proxy logic) and things like that.

If you are interested in any of the changes I've made, please merge them. If there's code that isn't very good, I'll update anything as required.

gartnera committed:
Removed short names from task list. Only display point values

gartnera committed:
Modified description text.

gartnera committed:
Updated task text.

gartnera committed:
Hide attatchment text if there is no attachment.

gartnera committed:
Modified the database structure.

Modified tasks and categories database layout to allow multiple tasks with same point value.
Added buildTables.sh to create tables (prevents table not found exceptions).
Modified the tasks retrieval and grid building code.
Modified the tasks template code.

gartnera committed:
Updated task route to accommodate database changes.

gartnera committed:
Removed unneeded enumerate.

gartnera committed:
Updated tasks code to show add buttons for categories and tasks.

gartnera committed:
Added admin_required decorator and associated error

gartnera committed:
Updated get_user code, calls, and template.

gartnera committed:
Fixed missing closing tag

gartnera committed:
Added functionaly to add categories

gartnera committed:
Moved admin check to grid build from template.

gartnera committed:
Added logic to allow add-box for empty categories

gartnera committed:
Added task creation functionality.

Also misc tweaks to fit new database layout.

gartnera committed:
Client side changes to match server side

gartnera committed:
Task addition template

gartnera committed:
Removed short_name from categories table.

gartnera committed:
Added percentage ring around tasks.

gartnera committed:
Added class to chall-box span

gartnera committed:
Added basic edit functionality.

Added sanitization to add and edit task functions.

gartnera committed:
Added edit button on task if admin.

gartnera committed:
Fixed if isAdmin wrong line

gartnera committed:
Added file upload functionality to task creation

gartnera committed:
Remove description <p></p> wrapping

gartnera committed:
Updated and added more SQL for delete cascade.

Supports deleting tasks.

gartnera committed:
Disable task flag input when solved.

gartnera committed:
Added logic for task page if no tasks present in database.

gartnera committed:
Added edit file (replace/add) file functionality

gartnera committed:
Initial work on mobile compatibility.

gartnera committed:
Added animated/interactive background. More mobile tweaks.

Background from https://github.com/nathanhleung/NodeCTF

gartnera committed:
Updated templates for new background

gartnera committed:
Updated SQL and added email field to users.

gartnera committed:
Added code to encforce sqlite foreign key constraits.

gartnera committed:
Added email field when logging in

gartnera committed:
Added proxy config option and associated code

gartnera committed:
Added IP logging

gartnera committed:
Changed font to improve cross platform compatibility.

gartnera committed:
Fixed error message not visible

gartnera committed:
Added configuration option for competition start time.

gartnera committed:
Update for uWSGI compatibility.

I'm not sure this is right, but it works.

gartnera committed:
Updated dependencies

gartnera committed:
Readded hidden user field. Hide admin by default.

gartnera committed:
Switched from htmllaundry to bleach

gartnera committed:
Updated deployment instructions

gartnera committed:
Added scoreboard.json url/endpoint

gartnera committed:
If no value specified in edit flag field, use the original value.

gartnera committed:
Removed old code and files.

gartnera committed:
Added command to create files folder

gartnera committed:
Updated image

gartnera committed:
Added delete button and functionality for tasks

gehaxelt commented:
@balidani @gartnera Nice! Is this project still actively maintained? 

I'm asking because I used it for a CTF a couple of days ago and would like to contribute some modifications (CSRF protection, start/finish time checks, ...) if @balidani is interested in them.

All the best,
gehaxelt

gartnera commented:
I'm around, but I don't think balidani is.

pjmc commented:
Hi, is there a way to delete a category? And to update the scoreboard? Thank you.

gartnera commented:
Not right now, I haven't gotten around to adding that. It's definitely on my todo list!

gartnera committed:
Simplified task grid page logic.

This will make deleting categories possible since there were indexing problems with the old approach.

gartnera committed:
Added category edit/delete functionality.

Just click on the category

gartnera committed:
Exclude hidden users from percent complete calculation.

gartnera committed:
Update dateutil/dateparser usages to match upstream changes.

Fixes #1

gartnera committed:
Added placeholder text for flag edit

gartnera committed:
Updated dependencies

gartnera committed:
Optimized tasks page generation code.

gartnera committed:
Added settings page to change password and email.

gartnera committed:
Removed animated background

gartnera committed:
Added requirements.txt

gartnera committed:
Added settings template

gartnera committed:
Made description optional.

gartnera committed:
Changed settings password field type.
